### DISCUSSION TITLE
How to make Stern-Gerlach B field with magpylib 2.3

### DISCUSSION THREAD

**Discussion by jesusmontera** [magpylib/magpylib#594] [Q&A] [answered]
I'm trying to calculate the non uniform Z (up) magnetic field for a space box 64x64x64 points (2 AU length). The particle will move from left to right  (x axis) in this box. I've made a **cylinder** rotated at top of the space and a **box** at the bottom like this:

```python
mup=magpy.source.magnet.Cylinder( mag = [0, B0*Bdir,0], dim = [Lmm*3,Lmm*3], pos=[0, 0, Lmm*2.7],  angle=-90,axis=[1,0,0])
                             
mdown=magpy.source.magnet.Box(mag=[0, 0,-B0*Bdir], dim=[Lmm*3, Lmm*3, Lmm*2], pos=[0, 0, -Lmm*2.2],angle=180,axis=[1,0,0])
```
Then I fill the B[x][y][z] np.array with collection  `getB([x,y,z])` for all space points.
Do a cylinder and a box will generate a non uniform field?
Is there a way to make B faster or better?

**Comment by OrtnerMichael** [comment]
hi @jesusmontera 

First of all: Yes, the field of a Cylinder Magnet as well as the field of a Cuboid magnet is inhomogenous. You can see this also nicely in an example given in the documentation:

https://magpylib.readthedocs.io/en/latest/_pages/page_01_introduction.html#intro-field-computation

Example 2 shows the field of a Cuboid.

![image](https://user-images.githubusercontent.com/47384012/223965061-c4d1502d-7a06-4e5c-b161-823194fb9344.png)

How to make your computation fast:

```python
import numpy as np
import magpylib as magpy

steps = 20
boxsize = 100

# create cube wiht path
cube = magpy.magnet.Cuboid(
    magnetization=(0,0,1),
    dimension=(1,1,1)
)
cube.position = np.linspace((-boxsize/2,0,0), (boxsize/2,0,0), steps)

# create grid
ts = np.linspace(-boxsize/2, boxsize, 64)
grid = np.array([[[(x,y,z) for x in ts] for y in ts] for z in ts])

# compute field
B = cube.getB(grid)
```

This takes a few seconds on my laptop.

Consider making use of the symmetry if you want to compute faster !
1. you only need to compute the field one time in a larger (wider) grid and then slice the respective 64x64x64 parts for you magnet position of interest. (increase speed by a lot)
2. you only need to compute the field in the first Octant - all other parts are mirrors thereof (increase speed by factor of 8)
3. If this is still not enough - make use of the core computation

does this answer your question ?

**Reply by jesusmontera** [reply]
Thanks but in version 2.3 (for python 3.6) there isn't a Cuboid shape, and replacing it in your sample with Cylinder, the program crashes on **getB** . Also in Stern-Gerlach aparatous there are two magnets, so I use a collection. and in my attempt (first  comment) that field with 2500 militeslas did not work well and took two hours. I will try rotating the cylinder  90º.

**Reply by Alexboiboi** [reply]
Hi @jesusmontera, If I may, a few questions:

1. Can you explain why you want to use a Cylinder and a Cuboid to generate your magnet field?
2. Is python 3.6 really a hard limitation of your environment? **A lot** has changed since v2 (currently 4.2, python 3.7+)!
3. I see that you want to read the field on a 64x64x64 grid. Are that many points helpful or even using a grid really relevant?  To solve equations of motions, it may not.
4. Are you exporting the result to continue your calculation with another tool?

More generally, If you would give us more substantial context about what you want to achieve, we may be able to provide better answers on how to help with your problem ;)

**Comment by jesusmontera** [comment]
Hello, my answers are:
1) I've tried with a cylinder and a box because is a similar shape to Stern-Gerlach configuration, but it didn't work. It worked better(but not perfect) with one box rotated at the top and 2 boxes rotated at the bottom(and only took 5 minutes)[ like this](https://github.com/jesusmontera/DiracMagnetic/blob/949e6df8e3d9bc23d77a5710bd6715ab97572a2c/magpylibgerlach/gerlach.py#L83).
2) I don't wan't to change of python version because  have lot's of packages that will take me long time to update.
3) The grid is relevant because the arrays to apply B  to the psi's must be that size, in [dot](https://github.com/jesusmontera/DiracMagnetic/blob/949e6df8e3d9bc23d77a5710bd6715ab97572a2c/auxfunctions.py#L70) or [potential](https://github.com/jesusmontera/DiracMagnetic/blob/949e6df8e3d9bc23d77a5710bd6715ab97572a2c/auxfunctions.py#L9) mode.
4)Yes,  the array is exported as npy file to [this app](https://github.com/jesusmontera/DiracMagnetic) to study how magnetic field acts on particle spin.

**Reply by Alexboiboi** [reply]
Hi, a few remarks to yours

1. I looked a bit at the code  in your repo, and there is plenty of optimization potential (also in the non-magpylib related code ;) ). The `getB` function provided by magpylib is only performant when using vectorized code, as is numpy. If you loop 64^3 times over it, no wonder the computation is very slow.
2. I don't see any reason not to upgrade at least to python 3.7 if you have, like it seems to be, control over your environment.  Python 3.6 has reached end of life over 1 year ago and except in some really really rare case there could be some breaking changes. This means that upgrading should not represent any effort from your side. The only work here would be to upgrade your magpylib code. This I am ready to help you with ;)
3. Ok if it used downstream

PS: I will post a solution later below in a new answer, stay tuned

**Comment by Alexboiboi** [comment, marked as answer]
Hi @jesusmontera,

Below I wrote a version based on your `gerlach.py` file, that takes advantage of vectorized code as described above by @OrtnerMichael. Note that I didn't not consider the validity of the 3-Cuboid system as a  Stern-Gerlach approximation, maybe @OrtnerMichael  can tell more about this choice, also because of the 1e-8mm cuboid side length?

The following example has been written with magpylib v4.2.0.

```python
# create a Stern-Gerlach B-field approximation with 3 cuboid magnets
from itertools import product

import magpylib as magpy
import numpy as np

# define constants
N = 64 # number of grid points per side
Bdir = -1  # Field direction
L_au = 2.0  # space side length in atomic units
a0_mm = 5.29177210903e-8  ## Bohr length in millimeters
L_mm = L_au * a0_mm  # space side length in millimeters
B0 = 1500.0  # militeslas
B1 = B0 / (2 ** 0.5)

# create magnetic system
mag_dim = [L_mm * 6, L_mm * 2, L_mm * 2]  # cuboids dimensions
mag_upper = magpy.magnet.Cuboid(
    magnetization=[0, -B1, -B1],
    dimension=mag_dim,
    position=[0, 0, L_mm * 2],
)
mag_upper.rotate_from_angax(45, [1, 0, 0])

mag_lower1 = magpy.magnet.Cuboid(
    magnetization=[0, B1, B1],
    dimension=mag_dim,
    position=[0, L_mm * 1.4, -L_mm * 2],
)
mag_lower1.rotate_from_angax(45, [1, 0, 0])
mag_lower2 = magpy.magnet.Cuboid(
    magnetization=[0, B1, B1],
    dimension=mag_dim,
    position=[0, -L_mm * 1.4, -L_mm * 2],
)
mag_lower2.rotate_from_angax(45, [1, 0, 0])

coll = magpy.Collection(
    mag_upper, mag_lower1, mag_lower2, style_label="Stern-Gerlach approximation"
)

# add object to visualize points grid space.
sens = magpy.Sensor(
    pixel=np.array(list(product(*[[-1, 1]] * 3))) * L_mm / 2,
    style_label="Grid space limits",
    style_pixel_size=0.5,
)

# flip system arount x-axis if field direction `Bdir` is negative
if Bdir == -1:
    coll.rotate_from_angax(180, [1, 0, 0], anchor=(0, 0, 0))

# display system
magpy.show(coll, sens, style_magnetization_size=0.3)

# compute B_array
ls = np.linspace(
    -L_mm / 2, L_mm / 2, N, endpoint=False
)  # add endpoint=False if you want same behavior as in your example
points = np.array([[x, y, z] for x in ls for y in ls for z in ls])
B_array = coll.getB(points).reshape(N, N, N, 3)
```

<img width="386" alt="image" src="https://user-images.githubusercontent.com/29252289/224508785-665e1c42-1dff-4341-9eb6-d86d09b1e2b0.png">

Here the timing on my old local machine (far from 5min ^^)
<img width="493" alt="image" src="https://user-images.githubusercontent.com/29252289/224508809-a990f7a6-bce1-4cf6-93da-5a6ab7dd5d0f.png">
Note that replace the source by cylinder will make the computation time only around twice as slow.

PS: I also noticed in your example that your points grid do not reach the positive grid space half in x,y or z direction (`+Lmmhalf` in your code). In the example above I included this point. If you want your old behavior add the `endpoint=False` argument to the `np.linspace` call as mentioned in the code comment.

PS2: I wrote an equivalent version also to be compatible with magpylib v2.3.0b0 but I ran into a `ValueError` exception when running `getB` on the grid points, but it is not planed to make a bug fix for such a old version...


Hope this will motivate you to upgrade to the newest magpylib version and enjoy the newest features ;)

**Comment by jesusmontera** [comment]
Thanks, I've converted your code to V 2.3 changing cuboid to box, and because it only take's a few seconds I changed the app, so the field can be created from the magnetic window rather than running an additional  script.
